File: go_site/com.gemaestudio/domain/Project.go
package domain

import (
	"database/sql"
	"log"
)

type ProjectImage struct {
	Id      int
	Image   string
	Project int
}

type PortfolioType struct {
	Id   int
	Name string
}

type Project struct {
	Id                    int
	PortfolioType         int
	Name                  string
	Image                 string
	EnableMainPageDisplay bool
	Description           string
	MainPageDescription   string
	PresentationOrder     int
	Gallery               []*ProjectImage
}

func (project *Project) findAll(con *sql.DB, sqlQuery string) (fullListing []*Project) {
	if sqlQuery == "" {
		sqlQuery = "SELECT * FROM project"
	}

	rows, err := con.Query(sqlQuery)
	if err != nil { /* error handling */
		log.Fatalf("Error on executing query: %s", err.Error())
	}

	fullListing = make([]*Project, 0, 10)
	for rows.Next() {
		exC := new(Project)
		err = rows.Scan(&exC.Id, &exC.Name, &exC.Image, &exC.PresentationOrder, &exC.PortfolioType, &exC.EnableMainPageDisplay, &exC.Description, &exC.MainPageDescription)
		if err != nil { /* error handling */
			log.Println("Error on scanning Project: %s", err.Error())
		}
		fullListing = append(fullListing, exC)
	}

	return
}

func (project *Project) FindAllMainPage(con *sql.DB) (listProjects []*Project) {
	return project.findAll(con, "SELECT * FROM project where enable_main_page_display = 1 order by presentation_order")
}

func (project *Project) FindByPortfolioType(con *sql.DB, id int) (fullListing []*Project) {
	rows, err := con.Query("SELECT * FROM project where portfolio_type = ?", id)
	if err != nil { /* error handling */
		log.Fatalf("Error on executing query: %s", err.Error())
	}

	fullListing = make([]*Project, 0, 7)
	for rows.Next() {
		exC := new(Project)
		err = rows.Scan(&exC.Id, &exC.Name, &exC.Image, &exC.PresentationOrder, &exC.PortfolioType, &exC.EnableMainPageDisplay, &exC.Description, &exC.MainPageDescription)
		if err != nil { /* error handling */
			log.Println("Error on scanning Project: %s", err.Error())
		}
		exC.Gallery = project.findGallery(con, exC.Id)
		fullListing = append(fullListing, exC)
	}

	return
}

func (project *Project) FindById(con *sql.DB, id int) (exC *Project) {
	exC = new(Project)

	// Prepare statement for reading data
	stmtOut, err := con.Prepare("SELECT * FROM project where id = ?")
	if err != nil {
		log.Fatalf("Error on executing query: %s", err.Error())
	}
	defer stmtOut.Close()

	err = stmtOut.QueryRow(id).Scan(&exC.Id, &exC.Name, &exC.Image, &exC.PresentationOrder, &exC.PortfolioType, &exC.EnableMainPageDisplay, &exC.Description, &exC.MainPageDescription) // WHERE number = 13
	if err != nil {
		log.Println("Error on scanning (by id) a project: %s", err.Error())
	}

	exC.Gallery = project.findGallery(con, exC.Id)

	return
}

func (project *Project) findGallery(con *sql.DB, projectId int) (fullListing []*ProjectImage) {
	rows, err := con.Query("SELECT * FROM project_images where project = ? order by id desc", projectId)
	if err != nil { /* error handling */
		log.Fatalf("Error on executing query: %s", err.Error())
	}

	fullListing = make([]*ProjectImage, 0, 2)
	for rows.Next() {
		exC := new(ProjectImage)
		err = rows.Scan(&exC.Id, &exC.Image, &exC.Project)
		if err != nil { /* error handling */
			log.Println("Error on scanning ProjectImage: %s", err.Error())
		}
		fullListing = append(fullListing, exC)
	}

	return
}

func (project *Project) FindAllPortfolioTypes(con *sql.DB) (fullListing []*PortfolioType) {
	rows, err := con.Query("Select * from portfolio_type where id in (select distinct(portfolio_type) from project)")

	if err != nil { /* error handling */
		log.Fatalf("Error on executing query: %s", err.Error())
	}

	fullListing = make([]*PortfolioType, 0, 3)
	for rows.Next() {
		exC := new(PortfolioType)
		err = rows.Scan(&exC.Id, &exC.Name)
		if err != nil { /* error handling */
			log.Println("Error on scanning PortfolioType: %s", err.Error())
		}
		fullListing = append(fullListing, exC)
	}

	return
}

File: scala_play_site/app/com/gemaestudio/Project.scala
package com.gemaestudio

import anorm._
import play.api.db._
import play.api.Play.current


class ProjectImage(var id: Pk[Long] = NotAssigned, val image: String, val project: Option[Int]) {
	override def toString = 
	  "prjImg: { id:"+id +
	  ", image:"+image+
	  ", project:"+project+"}"
}

class ProjectType(var id: Pk[Long] = NotAssigned, val name: String) {
	override def toString = 
	  "prjType: { id:"+id +
	  ", name:"+name+"}"
}

class Project(var id: Pk[Long] = NotAssigned, val portfolio_type: Option[Int], val name: String, val image: String, val enable_main_page_display: Boolean = false, 
    		val description: Option[String], val main_page_description: Option[String], val presentation_order: Option[Int]) {
  
    var gallery = List[ProjectImage]();

    override def toString = 
	  "{id:'"+id +
	  "', portfolio_type:'"+portfolio_type+
	  "', name:'"+name+
	  "', image:'"+image+
	  "', enable_main_page_display:'"+enable_main_page_display+
	  "', description:'"+description+
	  "', main_page_description:'"+main_page_description+
	  "', presentation_order:'"+presentation_order+
	  "', gallery: '"+
	  		gallery +
	  "'}"

	def add(img: ProjectImage) = gallery ::= img
}


object Project {
  def apply(id: Pk[Long] = NotAssigned, portfolio_type: Option[Int], name: String, image: String, enable_main_page_display: Boolean = false, 
    		description: Option[String], main_page_description: Option[String], presentation_order: Option[Int]) =
    new Project(
      id = id,
      portfolio_type = portfolio_type,
      name = name,
      image = image,
      enable_main_page_display = enable_main_page_display,
      description = description,
      main_page_description = main_page_description,
      presentation_order = presentation_order)

  def findByPortfolioType( id: Long ) : List[Project] = {
	import anorm.SqlParser.{get,str}
	import anorm.Pk
	 
	// Create a row parser object
	val rp: RowParser[Pk[Long]~Option[Int]~String~String~Boolean~Option[String]~Option[String]~Option[Int]] = get[Pk[Long]]("id") ~ get[Option[Int]]("portfolio_type") ~ str("name") ~ str("image") ~ get[Boolean]("enable_main_page_display") ~ get[Option[String]]("description") ~ get[Option[String]]("main_page_description") ~ get[Option[Int]]("presentation_order")
	val results: List[Project] = DB.withConnection("default") { implicit connection =>
	    // Parse All Results from SQL Statement
	    val rsp : ResultSetParser[List[Pk[Long]~Option[Int]~String~String~Boolean~Option[String]~Option[String]~Option[Int]]] = rp *
	    val results = SQL("SELECT * FROM project where portfolio_type = {id}") on("id" -> id) as ( rsp )
	    // Use pattern matching to re-construct our objects
	    results map {
	        case pk~portfolio_type~name~image~enable_main_page_display~description~main_page_description~presentation_order  => Project(pk,portfolio_type,name,image,enable_main_page_display,description,main_page_description,presentation_order)
	    }
	}
	results.foreach(proj => findGallery(proj.id).foreach(img => proj.add(img)))
	results
  }

  def findGallery( projectId: Pk[Long] ) : List[ProjectImage] = {
	import anorm.SqlParser.{get,str}
	import anorm.Pk
	 
	// Create a row parser object
	val rp: RowParser[Pk[Long]~String~Option[Int]] = get[Pk[Long]]("id") ~ str("image") ~ get[Option[Int]]("project")
	DB.withConnection("default") { implicit connection =>
	    // Parse All Results from SQL Statement
	    val rsp : ResultSetParser[List[Pk[Long]~String~Option[Int]]] = rp *
	    val results = SQL("SELECT * FROM project_images where project = {id} order by id desc") on("id" -> projectId) as ( rsp )
	    // Use pattern matching to re-construct our objects
	    results map {
	        case pk~image~project => new ProjectImage(pk,image,project)
	    }
	}
  }

  def findAll(sqlQuery: String = "SELECT * FROM project") : List[Project] = {
	import anorm.SqlParser.{get,str}
	import anorm.Pk
	 
	// Create a row parser object
	val rp: RowParser[Pk[Long]~Option[Int]~String~String~Boolean~Option[String]~Option[String]~Option[Int]] = get[Pk[Long]]("id") ~ get[Option[Int]]("portfolio_type") ~ str("name") ~ str("image") ~ get[Boolean]("enable_main_page_display") ~ get[Option[String]]("description") ~ get[Option[String]]("main_page_description") ~ get[Option[Int]]("presentation_order")
	val results: List[Project] = DB.withConnection("default") { implicit connection =>
	    // Parse All Results from SQL Statement
	    val rsp : ResultSetParser[List[Pk[Long]~Option[Int]~String~String~Boolean~Option[String]~Option[String]~Option[Int]]] = rp *
	    val results = SQL(sqlQuery) as ( rsp )
	    // Use pattern matching to re-construct our objects
	    results map {
	        case pk~portfolio_type~name~image~enable_main_page_display~description~main_page_description~presentation_order  => Project(pk,portfolio_type,name,image,enable_main_page_display,description,main_page_description,presentation_order)
	    }
	}
	results.foreach(proj => findGallery(proj.id).foreach(img => proj.add(img)))
	results
  }

  def findAllMainPage() : List[Project] = {
    findAll("SELECT * FROM project where enable_main_page_display = 1 order by presentation_order")
  }

  def findById(id: Option[Long]) : Project = {
	import anorm.SqlParser.{get,str}
	import anorm.Pk
	 
	// Create a row parser object
	val rp: RowParser[Pk[Long]~Option[Int]~String~String~Boolean~Option[String]~Option[String]~Option[Int]] = get[Pk[Long]]("id") ~ get[Option[Int]]("portfolio_type") ~ str("name") ~ str("image") ~ get[Boolean]("enable_main_page_display") ~ get[Option[String]]("description") ~ get[Option[String]]("main_page_description") ~ get[Option[Int]]("presentation_order")
	val results: List[Project] = DB.withConnection("default") { implicit connection =>
	    // Parse All Results from SQL Statement
	    val rsp : ResultSetParser[List[Pk[Long]~Option[Int]~String~String~Boolean~Option[String]~Option[String]~Option[Int]]] = rp *
	    val results = SQL("SELECT * FROM project where id = {id}") on("id" -> id) as ( rsp )
	    // Use pattern matching to re-construct our objects
	    results map {
	        case pk~portfolio_type~name~image~enable_main_page_display~description~main_page_description~presentation_order  => Project(pk,portfolio_type,name,image,enable_main_page_display,description,main_page_description,presentation_order)
	    }
	}
	results.foreach(proj => findGallery(proj.id).foreach(img => proj.add(img)))
	results(0)
  }
  
  def findExistingProjectTypes() : List[ProjectType] = {
	import anorm.SqlParser.{get,str}
	import anorm.Pk
	 
	// Create a row parser object
	val rp: RowParser[Pk[Long]~String] = get[Pk[Long]]("id") ~ str("name")
	DB.withConnection("default") { implicit connection =>
	    // Parse All Results from SQL Statement
	    val rsp : ResultSetParser[List[Pk[Long]~String]] = rp *
	    val results = SQL("select * from portfolio_type where id in (select distinct(portfolio_type) from project)") as ( rsp )
	    // Use pattern matching to re-construct our objects
	    results map {
	        case pk~name => new ProjectType(pk,name)
	    }
	}
  }
}